### PR TITLE
Some minor changes to permissions checking and logging

### PR DESCRIPTION
**Pull request by timrae**
1) Allow use of provider by com.ichi2.anki via `checkCallingOrSelfPermission()`

The only difference between `checkCallingOrSelfPermission()` and `checkCallingPermission()` is that the latter has [this additional check](http://grepcode.com/file/repository.grepcode.com/java/ext/com.google.android/android/4.3_r2.1/android/app/ContextImpl.java#1549):

```java
int pid = Binder.getCallingPid();
if (pid != Process.myPid()) {
    return checkPermission(permission, pid, Binder.getCallingUid());
}
return PackageManager.PERMISSION_DENIED;
```

so using `checkCallingOrSelfPermission()` (which doesn't have the above check) is essentially doing the same as what was there before in a more reliable way.

My understanding of the warnings in the documentation against using `checkCallingOrSelfPermission()` are that (as I read [in this paper](https://www.cs.rice.edu/~ds20/PermissionFlow-JRD-2013.pdf)) the warnings in the docs mainly apply to Activities that can be publicly invoked and put sensitive information into their activity result. 

There doesn't appear to be any way that us using `checkCallingOrSelfPermission()` in the ContentProvider could cause such an issue, but following the advice in the paper, I've set the "exported" property to False in the NoteEditor to prevent any permission leaking from there, since it has an intent filter setup and returns some information in its result.

2) There is a way to get the calling package for the purpose of logging and manifest inspection on all APIs, with the caveat that it only works if the host app is not sharing its UID with another package (which is quite unlikely), so I've enabled that.

### CODE REVIEW

**Comment by marcardar** [comment]
I think it's still useful to check the whether the calling package is AnkiDroid, because someone a some point in the future, may unwittingly set the [multiprocess](http://developer.android.com/guide/topics/manifest/provider-element.html#multi) flag to true.

> Activities that can be publicly invoked

Don't those activities still run in the target app's process?

> since it has an intent filter setup and returns some information in its result.

Does the activity show or just return a result without showing? If it shows, then I don't see a problem returning data that way.

**Comment by timrae** [comment]
> I think it's still useful to check the whether the calling package is AnkiDroid, because someone at some point in the future, may unwittingly set the multiprocess flag to true.

Why do you think that other AnkiDroid processes wouldn't have READ_WRITE_PERMISSION? I assume they would.

> Don't those activities still run in the target app's process?

I don't think so, otherwise the paper that I linked to wouldn't make any sense

> If it shows, then I don't see a problem returning data that way.

Return results from our Activities are not designed to be shared with 3rd party applications. As per the paper, the combination of an intent filter and default `exported` setting leads to `exported` being true, which wasn't our intention. Therefore I have set `exported = false` to match the other Activities.

**Comment by marcardar** [comment]
> Why do you think that other AnkiDroid processes

When multiprocess="true" the provider runs in the client process, so there is no IPC call. I'm not sure if this affects what you are talking about, but the possibility is worth bearing in mind.

> I don't think so, otherwise the paper that I linked to wouldn't make any sense

There must be something I'm missing. How are those activities started?  Obviously if a client app calls startActivity() with and AnkiDroid intent, then that activity is not running in the client process. Or have I totally misunderstood the Android framework!?

> Return results from our Activities are not designed to be shared with 3rd party applications.

Yeah, definitely it makes sense to set export="false" wherever possible.  Having said that, your NoteEditor example gives me an idea - it would be very useful for client apps to be able to open a NoteEditor (passing in the noteId).

**Comment by timrae** [comment]
Hmm yeah you're right, it looks like I still didn't understand the issue properly at all... I guess we really need to look under the hood of `checkPermission()` to see how that works. The implementation is not so easy to find beneath all the delegates and interfaces though.

**Comment by timrae** [comment]
@flerda 

Maybe you have some insight? `checkCallingOrSelfPermission()` always returns false when called via IPC, yet the docs say:

> Determine whether the calling process of an IPC or you have been granted a particular permission. This is the same as checkCallingPermission(String), except it grants your own permissions if you are not currently processing an IPC. Use with care!

What is the danger here?

**Comment by timrae** [comment]
By the way, I had a go at adding URI permissions within the ContentProvider to allow editing / deleting / updating of notes / models that were inserted via the provider. Unfortunately it doesn't appear to work though... `grantUriPermission()` returns without failing, but `checkUriPermission()` continue to return -1.

Here's the branch incase anyone has any ideas:
https://github.com/timrae/Anki-Android/commits/uri-permissions

**Comment by timrae** [comment]
I couldn't find any satisfactory answers on Stackoverflow, so I added a bounty to an [existing unanswered question](http://stackoverflow.com/questions/26140013/exploiting-checkcallingorselfpermission-for-privlege-escalation-attack) on the topic. No bites yet though unfortunately (I also posted a link to it in the Android Developers Google+ group).

**Comment by flerda** [comment]
> Yet the docs have the following warning:
> 
> > Determine whether the calling process of an IPC or you have been granted a particular permission. This is the same as checkCallingPermission(String), except it grants your own permissions if you are not currently processing an IPC. Use with care!
> 
> What is the danger here?

This is *my understanding* of what happens and what can go wrong. Take it with a grain of salt.

Whenever another process requests something of your application that may return a value (e.g., bound service, content provider, activity started "for result), one of your components in executed (in your application) but this is serving the request of the caller so, ```checkCallingPermission()``` will check that the caller (as returned by ```Binder.getCallingUid()```, which applies to all cases of IPC) has the given permission. This is to prevent permission escalation: for example, if an application's content provider wants to check that the other party has the correct permission, it can use checkCallingPermission() to do so; calling ```checkCallingPermission()``` without a caller (i.e., from the process that exposes the content provider) would lead to the permission being considered denied. This is useful if you always expect the request to be done from another process. For example, for the AnkiDroid API, if the content provider should only be used by third party apps (because AnkiDroid itself uses SQLite database and libanki directly) then ```checkCallingPermission()``` should suffice.

One reason this is important is that it is possible to reset the calling process (```Binder. clearCallingIdentity()```, possibly then followed by ```Binder. restoreCallingIdentity()```); this is useful, for example, to execute code on behalf of the application itself, even if the caller does not have those permissions. As a typical example, one would first check the caller has the right permissions, validate the inputs, then reset the calling identify and perform a privileged operation (one which would fail with the calling identify set, but is allowed by the application process), and then restore the calling identity.

Using checkCallingOrSelfPermission() will actually succeed if the code is invoked without a caller being set (e.g., because it has been cleared) and the application can itself perform the operation, which might lead to escalation.

I think the warning is there because if you write code that you expect to be invoked by another process and it is not, you would fail the permission check, which is generally safer.

*In the multi-process case*, things are not necessarily better with either of them, because the calling process, while it is within the same application, it is not the same process so both methods should do the same thing and ```checkCallingOrSelfPermission()``` would not help there. In that case, it would be better to grant the permission requested of the caller to the application itself and then ```checkCallingPermission()``` would work correctly.

In terms of exported activities, the best is always to not export activities, because they are possible attack vectors, even when they do not return a result. My preference would be to always explicitly marked them as ```android:exported="false"``` or ```android:expected="true"```, preferring the former, and adding validation for the latter.

I hope it helps.

**Comment by marcardar** [comment]
> In terms of exported activities

I'm going off on a tangent here, but...

You might want to consider having an exported "RedirectActivity" that can act as a single entry point for 3rd party apps. You then redirect to the appropriate non-exported activity. This has obvious advantages regarding keeping most of your activities non-exported as well as freedom to refactor without breaking 3rd party clients. I mention this now because I would like a "delete empty cards" intent so that my app can take the user to the AnkiDroid app and immediately show them the delete-empty-cards confirmation dialog.

Another use for the RedirectActivity might be for requesting user grants AnkiDroid storage permission (Android M)

For the following, you could just use uris (maybe already implemented?):

1. Show a note (possibly in edit mode)
2. Show a card or a card type
3. Pick a deck (uri/id returned in onActivityResult)
4. Pick a note (uri/id returned in onActivityResult)

**Comment by timrae** [comment]
> Using checkCallingOrSelfPermission() will actually succeed if the code is invoked without a caller being set (e.g., because it has been cleared) and the application can itself perform the operation, which might lead to escalation.

> I think the warning is there because if you write code that you expect to be invoked by another process and it is not, you would fail the permission check, which is generally safer.

Thanks a lot @flerda, I was able to confirm that indeed `checkCallingOrSelfPermission()` will always pass after a call to `Binder.clearCallingIdentity()`, as that will force `Binder.getCallingPid()` to have the same process as us.

It's good to know that there is unlikely to be any issue here, as we don't use this method, but I agree that it's a good practice to only allow the access that's necessary. In this case I have opted to use `getCallingOrSelfPermission()` on debug builds (to allow unit tests to pass), and the more restrictive `getCallingPermission()` for release builds.

> My preference would be to always explicitly marked them as android:exported="false"

Agreed, I have now made all the Activities other than the IntentHandler explicitly not exported

**Comment by marcardar** [comment]
> I was able to confirm that indeed checkCallingOrSelfPermission() will always pass after a call to Binder.clearCallingIdentity()

Please can you also check what happens when `multiprocess` is set to true?

**Comment by timrae** [comment]
>  I have now made all the Activities other than the IntentHandler explicitly not exported

Actually when I just checked, setting exported=false on the NoteEditor prevents the Share intent from working with it. I guess we'd better keep it exported.

**Comment by timrae** [comment]
> Please can you also check what happens when multiprocess is set to true?

As expected `checkCallingPermission()` and `checkCallingOrSelfPermission()` both continue to pass when multi-process is set to true.

**Comment by marcardar** [comment]
> As expected checkCallingPermission() and checkCallingOrSelfPermission() both continue to pass when multi-process is set to true.

Thanks - did you check both cases where client app has and does not have permission?

**Comment by timrae** [comment]
Sorry I don't really have time to investigate a flag that we don't use... and I expect it to fail for both as expected.

**Comment by marcardar** [comment]
No worries. I ran my own tests with:

1. permission granted/denied on client app
2. multiprocess true/false

And for all 4 combos I got the expected results :)

`checkCallingOrSelfPermission()` and `checkCallingPermission()` always returned same value.

**Comment by marcardar** [comment]
Hmm, on closer inspection of multiprocess it looks like it is a noop.  I checked Binder.getCallingPid() and it's still different from Process.myPid(). So that would explain why it's giving the same results. Over 5 years ago, Dianne Hackborn said:

> [Don't use this, it is some old cruft from pre-1.0 design that doesn't work and should be ignored these days.  Just pretend like the attribute doesn't exist. :}](https://groups.google.com/forum/#!topic/android-developers/u9UMJtALSXw)

**Comment by marcardar** [comment]
One important point to note in case AnkiDroid ever does use its own contentprovider in future is that checkCallingOrSelfPermission() returns DENIED if the user has not yet granted "com.ichi2.anki.permission.READ_WRITE_DATABASE" permission to AnkiDroid (note: getCallingPermission() always returns DENIED - as expected because not IPC).

So to be clear: on Android M, AnkiDroid will not be granted its own permission (<permission> tag in manifest) until the user explicitly grants it.

The bottom line: checkCallingOrSelfPermission() does what you want, so long as AnkiDroid is running on pre-M OR the user has explicitly granted "com.ichi2.anki.permission.READ_WRITE_DATABASE" to AnkiDroid.
